src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,683 @@
+{
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    using TypeParameterResolutions = ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    public static class TypeParamStuff
+    {
+        // TODO:
+        // This is the method that should be invoked to verify cycles of interest,
+        // i.e. where each callable in the cycle is type parametrized.
+        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined.
+        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        {
+            var parent = rootNode.CallableName;
+            var validResolution = TryCombineTypeResolutions(parent, out var combined, edges.Select(edge => edge.ParamResolutions).ToArray());
+            var resolvedToConcrete = combined.Values.All(res => !(res.Resolution is ResolvedTypeKind.TypeParameter tp) || tp.Item.Origin.Equals(parent));
+            return validResolution && resolvedToConcrete;
+            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
+            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+        }
+
+        public static TypeParameterResolutions Other(TypeParameterResolutions input)
+        {
+            static Tuple<QsQualifiedName, NonNullable<string>> AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+
+            var output = input.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            var mayBeReplaced = output
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+            var replacingParams = input.Where(kvp => mayBeReplaced.Contains(kvp.Key));
+            foreach (var (typeParam, paramRes) in replacingParams)
+            {
+                foreach (var keyInCombined in mayBeReplaced[typeParam])
+                {
+                    output[keyInCombined] = paramRes;
+                }
+            }
+
+            return output.ToImmutableDictionary();
+        }
+
+        public static bool TryCombineTypeResolutions(QsQualifiedName target, out TypeParameterResolutions combined, params TypeParameterResolutions[] resolutions)
+        {
+            if (target == null) throw new ArgumentNullException(nameof(target));
+            var success = Temp(out combined, resolutions);
+            combined = combined.Where(kvp => kvp.Key.Item1.Equals(target)).ToImmutableDictionary();
+            return success;
+        }
+
+        public static bool Temp(out TypeParameterResolutions combined, params TypeParameterResolutions[] resolutionDictionaries)
```

Comment:
Need to have better name for `Temp`.

Reply:
Renamed this to `TryCombineTypeResolutions`, and the original `TryCombineTypeResolutions` is now `TryCombineTypeResolutionsWithTarget`.